metricbeat/module/kafka/broker.go

```diff
@@ -290,6 +303,114 @@ func (b *Broker) FetchPartitionOffsetFromTheLeader(topic string, partitionID int
 	return offset, nil
 }
 
+// FetchPartitionOffsetsForTopics fetches offsets of all partitions from its leaders.
+func (b *Broker) FetchPartitionOffsetsForTopics(topics []*sarama.TopicMetadata, time int64) map[string]map[int32]PartitionOffsets {
+	leaderTopicPartition := b.groupBrokersPerTopicPartitions(topics)
+	topicPartitionPartitionOffsets := b.fetchGroupedPartitionOffsetsPerBroker(leaderTopicPartition, time)
+	return topicPartitionPartitionOffsets
+}
+
+func (b *Broker) groupBrokersPerTopicPartitions(topics []*sarama.TopicMetadata) brokerTopicPartitionsMap {
+	leaderTopicPartition := brokerTopicPartitionsMap{}
+	for _, topic := range topics {
+		for _, partition := range topic.Partitions {
+			if _, ok := leaderTopicPartition[partition.Leader]; !ok {
+				leaderTopicPartition[partition.Leader] = []topicPartition{}
+			}
+			leaderTopicPartition[partition.Leader] = append(leaderTopicPartition[partition.Leader],
+				topicPartition{topic: topic.Name, partition: partition.ID})
+		}
+	}
+	return leaderTopicPartition
+}
+
+func (b *Broker) fetchGroupedPartitionOffsetsPerBroker(leaderTopicPartition brokerTopicPartitionsMap,
+	time int64) map[string]map[int32]PartitionOffsets {
+
+	var wg sync.WaitGroup
+	wg.Add(len(leaderTopicPartition))
+
+	queryResults := make(chan map[string]map[int32]PartitionOffsets, len(leaderTopicPartition))
+	defer close(queryResults)
+
+	for leader, topicPartitions := range leaderTopicPartition {
+		thisLeader := leader
+		thisTopicPartitions := topicPartitions
+		go func() {
+			queryResults <- b.queryBrokerForPartitionOffsets(thisLeader, thisTopicPartitions, time)
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+
+	topicPartitionPartitionOffsets := map[string]map[int32]PartitionOffsets{}
+	for i := 0; i < len(leaderTopicPartition); i++ {
+		queryResult := <-queryResults
+		for topic := range queryResult {
+			topicPartitionPartitionOffsets[topic] = queryResult[topic]
+		}
+	}
+	return topicPartitionPartitionOffsets
+}
+
+func (b *Broker) queryBrokerForPartitionOffsets(brokerID int32, topicPartitions []topicPartition, time int64) map[string]map[int32]PartitionOffsets {
+	req := new(sarama.OffsetRequest)
+	for _, topicPartition := range topicPartitions {
+		req.AddBlock(topicPartition.topic, topicPartition.partition, time, 1)
+	}
+
+	broker, err := b.client.Leader(topicPartitions[0].topic, topicPartitions[0].partition)
+	if err != nil {
+		return b.handleBrokerQueryError(topicPartitions, err)
+	}
+
+	defer func() {
+		err := broker.Close()
+		if err != nil {
+			log.Debugf("closing broker (ID: %d) failed: ", brokerID, err)
+		}
+	}()
+
+	resp, err := broker.GetAvailableOffsets(req)
+	if err != nil {
+		err = fmt.Errorf("get available offsets failed by leader (ID: %d): %v", brokerID, err)
```

Comment:
Nit. We could use `errors.Wrap` to add context to these errors and others added in this PR.